packages/javascript-decorators/src/memoize.ts

```diff
@@ -0,0 +1,51 @@
+export interface MemoizeMap<T, U> {
+  get(key: T): U;
+  has(key: T): boolean;
+  set(key: T, value: U): MemoizeMap<T, U>;
+}
+
+export const MAX_MAP_ENTIRES = 50;
+
+export default function Memoize(resolver?: Function) {
+  return function(...args: any[]) {
+    return Memoized(args[0], resolver);
+  };
+}
+
+function Memoized(func: Function, resolver?: Function) {
+  const weakMapCache = new WeakMap();
+  const mapCache = new Map();
+
+  return function(...args: any[]) {
+    if (typeof window === 'undefined') {
+      return func.apply(this, args);
+    }
+
+    const useWeakMap =
+      args.length === 0 && typeof args[0] === 'object' && !resolver;
+
+    let key;
+    if (useWeakMap) {
+      key = args[0];
+    } else if (resolver) {
+      key = resolver.apply(this, args);
```

Comment:
What is `this` here?

Reply:
I think this should be consistent with how the method is run.

Reply:
🐟 I can't think why we'd want the resolver to have access to the class instance.  That feels like a foot-gun.

Poking at this a bit: a resolver that tries to access `this.addOne` passes type checks, but fails at runtime because `this === undefined`.

Reply:
So this should probably be `key = resolver(...args)`?

Reply:
thank you for explaining! will change this.